[Port dspace-7_x] Cache-bust dynamic configuration files and theme CSS

ybnd opened the pull request:
## References
* Fixes #3992
* Fixes #1961

## Description
DSpace Angular's production server can modify the configuration it serves to CSR-only clients through YAML or environment variables. However, these files can remain cached in the browser and leave users with out-of-date configuration until the TTL runs out or the user does a "hard refresh". On build time this sort of problem is solved by saving the content hash as part of the path of each file (JS, CSS, ...)

We introduce HashedFileMapping to bridge the same gap for dynamic content generated _after_ the server is built:
- Files added to this mapping on the server are copied to a hashed path
- A copy is injected into index.html, where clients can read it out and resolve the hashed paths
- If a given path is not found in the mapping, the client will fall back to the original version (to prevent errors in development mode)

With this mechanism we can ensure updates to config.json and similar files take immediate effect without losing the performance benefit of client-side caching.

Compression support is included in anticipation of a pending PR for https://github.com/DSpace/dspace-angular/issues/3961, where the initial size of `config.json` can become large enough where compressing it would be a plus.

## Discussion
- It's not clear how/if we can unit test this at the time of writing
- Should we inlude a configuration flag to toggle this feature on/off? We may want to roll it out as opt-in so downstream users can wait it out until it has been battle-tested a bit.


## Instructions for Reviewers

To test:
1. Run DSpace Angular locally, in production mode
2. Confirm that you see the following files:
    - `[dist]/assets/config.<hash>.json` → hashed configuration file
    - `[dist]/dspace-theme.<hash>.css` → hashed theme CSS file
    - `[dist]/index.html` → index file with an extra `script#hashed-file-mapping` element containing the mapping of plain paths to hashed paths
3. Confirm that `config.<hash>.json` and `dspace-theme.<hash>.css` are correctly served to the browser, and that the app runs/looks as before the PR
5. Confirm that https://github.com/DSpace/dspace-angular/issues/3992 is resolved

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You do not need to complete this checklist prior creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

ybnd committed:
Cache-bust dynamic configuration files

DSpace Angular's production server can modify the configuration it serves to CSR-only clients through YAML or environment variables.
However, these files can remain cached in the browser and leave users with out-of-date configuration until the TTL runs out or the user does a "hard refresh".
On build time this sort of problem is solved by saving the content hash as part of the path of each file (JS, CSS, ...)

We introduce HashedFileMapping to bridge the same gap for dynamic content generated _after_ the server is built:
- Files added to this mapping on the server are copied to a hashed path
- A copy is injected into index.html, where clients can read it out and resolve the hashed paths
- If a given path is not found in the mapping, the client will fall back to the original version (to prevent errors in development mode)

With this mechanism we can ensure updates to config.json and similar files take immediate effect without losing the performance benefit of client-side caching.

ybnd committed:
Don't try to 'run' JSON on load

The JSON mapping needs to be declared as a data block, otherwise browers may complain (nothing seems to really break though, except for Cypress)
See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#any_other_value

ybnd committed:
Add theme CSS to hashed file map

ybnd commented:
@tdonohue adjusted the labels since this PR was opened against 7.x

ybnd committed:
Simplify testing by making mapping optional

Mainly for 8.x and beyond, where ThemeService is injected more often ~ standalone components

ybnd commented:
Note: when running multiple DSpace Angular instances in PM2 cluster mode this would cause the same filesystem operations in parallel, could cause problems.

Could be addressed with Redis similar to https://github.com/DSpace/dspace-angular/issues/3110#issuecomment-2188477157

ybnd committed:
Optionally preload hashed files

Both the configuration and theme CSS are known to be needed for every render.
We can save some time by retrieving them as soon as the HTML is read.

ybnd committed:
Don't hash or cache the index file

This didn't work well in one specific case: requests without a path (e.g. https://demo.dspace.org)
Such requests would result in a 304 redirect directly to index.html, losing the hashed file mapping and causing it to get cached.
Then it could remain in the cache across rebuilds.

Solutions:
- Don't try to hash index.html, but modify it in place
- Introduce configuration to disable caching for specific static files and apply this to index.html to prevent similar problems
- Don't let browsers cache index.html when it's served for CSR under another path

ybnd committed:
Merge remote-tracking branch 'origin/dspace-7_x' into cache-bust-dynamic-configuration-7.6

pcg-kk reviewed (commented):
Work for me fine, tested with SSR & without SSR

ybnd committed:
Fix one-letter variable names

ybnd committed:
Merge remote-tracking branch 'ybnd/cache-bust-dynamic-configuration-7.6_CLEAN' into cache-bust-dynamic-configuration-7.6

ybnd committed:
Small improvements

- Track head links in a Map, because Set doesn't deduplicate them properly
- Make ServerHashedFileMapping independent of ThemeConfig
- Make theme pre-fetching configurable (e.g. only prefetch the main theme CSS, but leave route-based theme CSS lazy)

ybnd committed:
Merge remote-tracking branch 'ybnd/cache-bust-dynamic-configuration-7.6_CLEAN' into cache-bust-dynamic-configuration-7.6

ybnd committed:
Merge remote-tracking branch 'origin/dspace-7_x' into cache-bust-dynamic-configuration-7.6

MW3000 reviewed (approved):
> Run DSpace Angular locally, in production mode
> Confirm that you see the following files:
> `[dist]/assets/config.<hash>.json` → hashed configuration file

Yes, `config.<hash>.json` files are correctly generated ✅
The `config.<hash>.json` is correctly referenced ✅

> `[dist]/dspace-theme.<hash>.css` → hashed theme CSS file

Yes, custom theme `theme.<hash>.css` files are correctly generated and referenced ✅

> `[dist]/index.<hash>.html` → index file with an extra script#hashed-file-mapping element containing the mapping of plain paths to hashed paths

There is no `index.<hash>.html` file generated, just the normal `index.html` file! But I guess, that is just an error in this description? As I understand the code, there should not be an `index.<hash>.html` file, as the normal `index.html` is being excluded from cashes? ❓

The `index.html` contains the correct hashed-file-mappings ✅

> Confirm that `config.<hash>.json` and `dspace-theme.<hash>.css` are correctly served to the browser, and that the app runs/looks as before the PR

Confirmed ✅

The feature works as described. I read the code and see what it does but I am not competent enough to decide if this is the right solution architecturally.

tdonohue reviewed (approved):
👍 Thanks @ybnd !  Verified this is the same as #4936.  I also see it's already been tested by others. Merging immediately.

This still needs to be ported to 9.x and 8.x but I see PRs for those already exist in draft form (see #4030 and #4031)
